File: embedding-service/src/modules/logger/logger.py
import json
import sys
from datetime import UTC, datetime


class Logger:
    def debug(self, message: str) -> None:
        self._write("debug", message)

    def info(self, message: str) -> None:
        self._write("info", message)

    def warn(self, message: str) -> None:
        self._write("warn", message)

    def error(self, message: str) -> None:
        self._write("error", message)

    def _write(self, level: str, message: str) -> None:
        entry = {
            "timestamp": datetime.now(UTC).isoformat(),
            "level": level,
            "message": message,
        }
        stream = sys.stderr if level in ("error", "warn") else sys.stdout
        print(json.dumps(entry), file=stream)


logger = Logger()

File: api-service/src/modules/logger/logger.ts
import type { ILogMeta, TLogLevel } from "@interface/log.interface.js";

export class Logger {
  debug(message: string, meta?: ILogMeta): void {
    this.write("debug", message, meta);
  }

  info(message: string, meta?: ILogMeta): void {
    this.write("info", message, meta);
  }

  warn(message: string, meta?: ILogMeta): void {
    this.write("warn", message, meta);
  }

  error(message: string, meta?: ILogMeta): void {
    this.write("error", message, meta);
  }

  private write(level: TLogLevel, message: string, meta?: ILogMeta): void {
    const entry = {
      timestamp: new Date().toISOString(),
      level,
      message,
      ...(meta ? { meta } : {}),
    };

    const serialized = JSON.stringify(entry);

    if (level === "error") {
      console.error(serialized);
      return;
    }

    if (level === "warn") {
      console.warn(serialized);
      return;
    }

    console.log(serialized);
  }
}

export const logger = new Logger();
